packages/core/src/components/axes/ruler.ts

```diff
@@ -0,0 +1,187 @@
+// Internal Imports
+import { Component } from "../component";
+import { DOMUtils } from "../../services";
+import { TooltipTypes, ScaleTypes } from "../../interfaces";
+
+// D3 Imports
+import { mouse, Selection } from "d3-selection";
+
+import { isEqual, flatten } from "lodash-es";
+
+type GenericSvgSelection = Selection<SVGElement, any, SVGElement, any>;
+
+const THRESHOLD = 5;
+
+/** check if x is inside threshold area extents  */
+function pointIsMatch(dx: number, x: number) {
+	return dx > x - THRESHOLD && dx < x + THRESHOLD;
+}
+
+export class Ruler extends Component {
+	type = "ruler";
+	backdrop: GenericSvgSelection;
+	hoveredElements: GenericSvgSelection;
+
+	render() {
+		const domainAxisPosition = this.services.cartesianScales.getDomainAxisPosition();
+		const domainScaleType = this.services.cartesianScales.getScaleTypeByPosition(
+			domainAxisPosition
+		);
+		const isTimeSeries = domainScaleType === ScaleTypes.TIME;
+
+		// if scale type is not timeSeries do not show rule
+		if (!isTimeSeries) {
+			return;
+		}
+
+		this.drawBackdrop();
+		this.addBackdropEventListeners();
+	}
+
+	showRuler([x, y]: [number, number]) {
+		const svg = this.parent;
+		const ruler = DOMUtils.appendOrSelect(svg, "g.ruler");
+		const line = DOMUtils.appendOrSelect(ruler, "line.ruler-line");
+		const dataPoints: GenericSvgSelection = svg.selectAll(
+			"[role=graphics-symbol]"
+		);
+		const { datasets } = this.model.getDisplayData();
+		const mainXScale = this.services.cartesianScales.getMainXScale();
+		const mainYScale = this.services.cartesianScales.getMainYScale();
+		const [yScaleEnd, yScaleStart] = mainYScale.range();
+
+		const scaledData: number[] = flatten(datasets.map(({ data }) =>
+			data.map((d, i) => Number(this.services.cartesianScales.getDomainValue(d, i)))
+		));
+
+		/**
+		 * Find matches, reduce is used instead of filter
+		 * to only get elements which belong to the same axis coordinate
+		 */
+		const scaledValuesMatches: number[] = scaledData.reduce((acc, cur) => {
+			const sampleAccValue = acc[0];
+
+			// if current value is bigger than already existing values forget it
+			if (sampleAccValue && cur > sampleAccValue) {
+				return acc;
+			}
+
+			// there's a match and cur is either less then or equal to already stored values
+			if (pointIsMatch(cur, x)) {
+				if (sampleAccValue && cur < sampleAccValue) {
+					// there's a closer data point in the threshold area, so reinstantiate array
+					acc = [cur];
+				} else {
+					// cur is equal to already stored values, there's another match on the same coordinate
+					acc.push(cur);
+				}
+			}
+
+			return acc;
+		}, []);
+
+		// some data point match
+		if (scaledValuesMatches.length > 0) {
```

Comment:
what does `scaledValuesMatches` do exactly? is there a better var name we can use?

Reply:
It's an array containing all the data points that match the current mouse x position, which are basically the ones that need to be triggered. I called it like that to be "extensible" (when we started working on this we planned to have the same ruler interaction on other charts with different graphic elements too) but if you have a better name we can change it, no prob 👍

Reply:
I think the name `dataPointsMatchingRulerLine` should be good, what do you think?